File: Java/src/main/java/org/codecop/dependencies/parameterise_constructor/Money.java
package org.codecop.dependencies.parameterise_constructor;

import java.math.BigDecimal;
import java.util.Objects;

public class Money {

    public static final Money ONE_THOUSAND = new Money(new BigDecimal("1000"));
    public static final Money ONE_HOUNDRED = new Money(new BigDecimal("100"));

    private final BigDecimal value;

    public Money(BigDecimal value) {
        this.value = value;
    }

    public Money reduceBy(int p) {
        return new Money(value.multiply(new BigDecimal(100 - p)).divide(new BigDecimal(100)));
    }

    public boolean moreThan(Money other) {
        return this.value.compareTo(other.value) > 0;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Money money = (Money) o;
        return Objects.equals(value, money.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    @Override
    public String toString() {
        return "Money{" +
                "value=" + value +
                '}';
    }
}

File: CSharp/DependencyBreakingKatas/ParameteriseConstructor/Money.cs
using System;

namespace Org.Codecop.Dependencies.ParameteriseConstructor
{
    public class Money
    {
        public static readonly Money OneThousand = new Money(1000);

        public static readonly Money OneHoundred = new Money(100);

        private readonly decimal value;

        public Money(int value) => this.value = value;

        public Money(decimal value) => this.value = value;

        public virtual Money ReduceBy(int p)
        {
            return new Money(value * (100m - p) / 100m);
        }

        public virtual bool MoreThan(Money other)
        {
            return this.value.CompareTo(other.value) > 0;
        }

        public override bool Equals(object other)
        {
            if (ReferenceEquals(this, other))
            {
                return true;
            }
            if (ReferenceEquals(null, other) || other.GetType() != this.GetType())
            {
                return false;
            }
            var that = (Money)other;
            return this.value.Equals(that.value);
        }

        public override int GetHashCode()
        {
            return value.GetHashCode();
        }

    }

}
